include/ceed/jit-source/magma/magma-basis-interp-deriv-nontensor.h

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+/// @file
+/// Internal header for MAGMA non-tensor basis interpolation
+#ifndef CEED_MAGMA_BASIS_INTERP_DERIV_NONTENSOR_H
+#define CEED_MAGMA_BASIS_INTERP_DERIV_NONTENSOR_H
+
+#include "magma-common-nontensor.h"
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T, int Q_COMP, int P, int Q, int NB>
+static __device__ __inline__ void magma_basis_nontensor_device_n(const int n, CeedScalar const *dA, CeedScalar const *dB, CeedScalar *dC,
+                                                                 CeedScalar *shared_data) {
+  const int tx      = threadIdx.x;
+  const int ty      = threadIdx.y;
+  const int id      = blockIdx.x * blockDim.y + ty;
+  const int nblocks = (n + NB - 1) / NB;
+  const int myn     = min(NB, n - id * NB);
+
+  dB += id * P * NB;
+  dC += id * Q * NB;
+
+  // A is P x Q
+  CeedScalar *sB = shared_data + ty * P * NB;
+  CeedScalar *sA = shared_data + blockDim.y * P * NB;
```

Comment:
`sB` is the first `P * NB * blockdim.Y` entries, where depending on the thread's y-index it points to a chunk of `P * NB` in that array.
`sA` points on all threads to the last `P * Q` block of shared memory.

It could just as easily be the opposite order, where the first `P * Q` entries are for `sA` and then after comes `sB`. Not sure if this has performance implications but as written it's in order of access.

Reply:
Ah, makes sense.